models/m_payment_mollie_api.erl

```diff
@@ -260,129 +443,299 @@ api_key(Context) ->
 
+mollie_list_subscriptions_next(#{ <<"next">> := Next }, Context) when is_binary(Next), Next =/= <<>> ->
+    case api_call(get, Next, [], Context) of
+        {ok, #{ <<"data">> := Data, <<"links">> := Links }} ->
+            case mollie_list_subscriptions_next(Links, Context) of
+                {ok, NextData} ->
+                    {ok, Data ++ NextData};
+                {error, _} = Error ->
+                    Error
+            end;
+        {error, _} = Error ->
+            Error
+    end;
+mollie_list_subscriptions_next(#{}, _Context) ->
+    {ok, []}.
+
+
+parse_sub(#{
+        <<"resource">> := <<"subscription">>,
+        <<"id">> := SubId,
+        <<"customerId">> := CustId,
+        <<"startDate">> := Start,
+        <<"cancelledDatetime">> := Cancelled,
+        <<"amount">> := Amount,
+        <<"interval">> := Interval,
+        <<"description">> := Description
+    }) ->
+    #{
+        id => {mollie_subscription, CustId, SubId},
+        is_valid => is_null(Cancelled),
+        start_date => to_date(Start),
+        end_date => to_date(Cancelled),
+        description => Description,
+        interval => Interval,
+        amount => z_convert:to_float(Amount)
+    }.
+
+is_null(undefined) -> true;
+is_null(null) -> true;
+is_null(<<>>) -> true;
+is_null(_) -> false.
+
+to_date(undefined) -> undefined;
+to_date(null) -> undefined;
+to_date(D) -> z_datetime:to_datetime(D).
+
+
 has_subscription(UserId, Context) ->
-    CustomerId = m_rsc:p(UserId, mollie_customer_id, z_acl:sudo(Context)),
-    case m_rsc:p(UserId, mollie_subscription_id, z_acl:sudo(Context)) of
-        undefined ->
-            false;
-        SubscriptionId ->
-        case api_call(get, "customers/" ++
-                          z_convert:to_list(CustomerId) ++
-                          "/subscriptions/" ++
-                          z_convert:to_list(SubscriptionId),
-                      [], Context) of
-            {ok, Response} ->
-                Status = maps:get(<<"status">>, Response),
-                Status == <<"active">>;
-            {error, 410}->
-                false;
-            {error, 404} ->
-                false;
-            {error, Error} ->
-                m_payment_log:log(
-                  CustomerId,
-                  "ERROR",
-                  [
-                   {psp_module, mod_payment_mollie},
-                   {description, "API Error fetching subscription info from Mollie"},
-                   {request_result, Error}
-                  ],
-                  Context),
-                lager:error("API error fetching mollie subscription for #~p: ~p", [CustomerId, Error]),
-                false
-        end
+    case list_subscriptions(UserId, Context) of
+        {ok, Subs} ->
+            lists:any(fun is_subscription_active/1, Subs);
+        {error, _} ->
+            false
     end.
 
+is_subscription_active(#{ end_date := undefined }) -> false;
+is_subscription_active(#{ end_date := _}) -> true.
```

Comment:
Shouldn't this be the other way round? If there is no end-date at which the sub was terminated, it should be considered active - if I'm correct I've seen that interpretation somewhere else in the code above.

Reply:
Good one!

Let me check this and the code calling this.

Reply:
You are right - I pushed a fix.